### ISSUE TITLE
NET-1985 New Rule S5770: View data dictionaries should be replaced by models

### ISSUE THREAD

**Issue by kidchenko** [SonarSource/sonar-dotnet#2602]
### ToDo:

Implement S5770

### Description

I would like to suggest a new rule. `An analyzer to point out the usage of ViewBag and replace it with an model`

### Expected behavior

MyModel.cs
````csharp

public string Message { get; set; }
````

MyController.cs
````csharp


model.Message = "Hello World;"


### Actual behavior

MyController.cs
````csharp

ViewBag.Message = "Hello World"; // no alert shown
````

### Why I think this is useful? 

ViewBag is a dynamic property and you will only catch errors at run time making your code error prone.

**Comment by andrei-epure-sonarsource** [comment]
hi @kidchenko . thanks for the suggestion - I agree with this rule. However, I am afraid it would be quite noisy, as many devs might just use ViewBag and the rule would not provide a lot of value. Or do you think that using ViewBag is rather a corner case?

**Comment by pavel-mikula-sonarsource** [comment]
ViewBag is easy to use + Using ViewBag is really bad idea = This rule can bring a lot of value.

While specifing, we should consider also ViewData and TempData as (configurable?) bad ideas.

cc @nicolas-harraudeau-sonarsource

**Comment by nicolas-harraudeau-sonarsource** [comment]
@andrei-epure-sonarsource @pavel-mikula-sonarsource `ViewBag` seems indeed to be used a lot ([Sourcegraph](https://sourcegraph.com/search?q=%22+ViewBag.%22+%22%3D%22+count:1000&patternType=regexp#627), [Github](https://github.com/search?l=&p=1&q=viewbag+extension%3Acs&type=Code)), including Microsoft projects.

If I understand correctly the main issue for this rule is that ViewBag is not type checked, thus it makes code maintenance more difficult. This would be a code smell rule. Is that it?

If Microsoft documentation mentions that *"some developers prefer to minimally or never use ViewData and ViewBag"* I guess a rule which is not in "Sonar way" Quality Profile would be possible, but it would have a low priority. My main questions are:
* Will everybody accept this rule and change their code? If yes then we can put this rule in "Sonar way" Quality Profile.
* If not, are there clear use cases where using a ViewBag is ok? If yes:
  * Can we make exceptions for them?
  * Would we be able to help code maintenance by detecting obvious mistakes? We could then create a Bug rule detecting when ViewBag is misused.
* Are there use cases where it is never ok to use a ViewBag? We could then raise issues in these cases.

If we can find a scope in which raising an issue is always ok we can make a great rule out of it.

**Comment by pavel-mikula-sonarsource** [comment]
I would say `Code smell`+`Major`+`SonarWay`.

Main problem is that if you do typo like this `ViewBag.Tilte = "Something";` you'll never find out. Usage in view `@ViewBag.Title` will emit empty string, no exception, no warning.

Using ViewBag is easy + simple => dangerous. 
It's acceptable for small/demo projects, but it's never OK 🙂 => SonarWay. You can always deactivate the rule, but the bigger project you have, the worse idea it is. There's no clear condition when it's good.

We can consider ignoring some well-known words and/or provide configuration for it (Title, Keywords, Description). I'd prefer to always raise, unless configured by user.

Obvious mistake detecting means go through all cshtml/vbhtml files, find all usages and report on never used. That would not be very precise, unless we build some view-tree for each controller. That would be another high-value bug detection rule, but we don't parse cshtml/vbhtml files yet.

**Comment by nicolas-harraudeau-sonarsource** [comment]
Thanks for this clear explanation @pavel-mikula-sonarsource. Let's go for the  Code smell+Major+SonarWay rule as you suggested. Could you specify it? I can review it. Or I'll specify it in a few weeks and ping you if I have any question.

We analyze cshtml in SonarHTML but there is no priority to do it in SonarDotnet as far as I know. There wouldn't be much value in it anyway if using ViewBag is a bad idea. The simpler the rule the better.

**Comment by kidchenko** [comment]
sure agreed! +1

**Comment by nicolas-harraudeau-sonarsource** [comment]
I was saying that to @pavel-mikula-sonarsource but if you want to specify it @kidchenko be my guest ;) Is this what your last message meant?

**Comment by pavel-mikula-sonarsource** [comment]
@nicolas-harraudeau-sonarsource I'll write the RSPEC and you'll do the review.

**Comment by kidchenko** [comment]
@nicolas-harraudeau-sonarsource @pavel-mikula-sonarsource I saw the work is in progress already, just created my account in JIRA (not sure if I will need)

**Comment by pavel-mikula-sonarsource** [comment]
Hello @kidchenko ,

Thank you one more time for this valuable suggestion. The rule is not specified as [RSPEC-5770](https://jira.sonarsource.com/browse/RSPEC-5770). There's no ETA to implement it, but specification is the first step on it's way to SonarWay profile.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1985](https://sonarsource.atlassian.net/browse/NET-1985)

[NET-1985]: https://sonarsource.atlassian.net/browse/NET-1985?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pavel-mikula-sonarsource** [comment]
Moved
